internal/sync2/keylock.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"sync"
+)
+
+// KeyLock provides per-key RW locking. Locking is key-specific, meaning lock
+// contention only exists for locking calls using the same key parameter. As
+// with all locks, do not call Unlock() or RUnlock() before a corresponding
+// Lock() or RLock() call.
+//
+// Note on memory usage: internally KeyLock lazily and atomically creates a
+// separate sync.RWMutex for each key. To maintain synchronization guarantees,
+// these interal mutexes are not freed until the entire KeyLock instance is
+// freed.
+type KeyLock struct {
+	locks sync.Map
```

Comment:
I'm not quite sure whether using sync.Map here is the best approach here. I think it fits the regular optimization criteria, but I'm not certain. Using a regular map would allow using `storj.NodeID` as the key, rather than interface{}.

Reply:
From the sync.map documentation:

```The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow```

This fits that case.

I chose to attempt a general solution here, which is why I am not using `storj.NodeID` explicitly as the key.

Reply:
Yeah, but using sync.Map happens together with using interfaces, which may negate the performance benefit and/or end up using more memory.

Reply:
I have changed the implementation to use a direct map with a sync.Mutex and placed it in the overlay package. Performance was slightly better, and Paul was right that there wasn't a clear need for a generalized solution.